jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1164,10 +1168,29 @@ public Method getDeclaredMethod(String name, Class<?>... parameterTypes) throws
  */
 @CallerSensitive
 public Class<?> getDeclaringClass() {
-	Class<?> declaringClass = getDeclaringClassImpl();
+	Class<?> declaringClass = null;
+	if (cachedDeclaringClass == null) {
+		if (cachedDeclaringClassOffset == -1) {
+			try {
+				Field cachedDeclaringClassField;
+				cachedDeclaringClassField = Class.class.getDeclaredField("cachedDeclaringClass"); //$NON-NLS-1$
+				cachedDeclaringClassOffset = getUnsafe().objectFieldOffset(cachedDeclaringClassField);
+			} catch (NoSuchFieldException e) {
+				throw newInternalError(e);
+			}
+		}
+		/*[IF Sidecar19-SE]
+		getUnsafe().putObjectRelease(this, cachedDeclaringClassOffset, getDeclaringClassImpl());
+		/*[ELSE]*/
+		getUnsafe().putOrderedObject(this, cachedDeclaringClassOffset, getDeclaringClassImpl());
+		/*[ENDIF]*/
```

Comment:
This `getUnsafe().put...` code can be moved within `if (cachedDeclaringClassOffset == -1) {` to avoid being invoked repeatedly for those classes with `null declaringClass`.

Reply:
could synchronization problem happens here that another thread initializes cachedDeclaringClassOffset but not setting cachedDeclaringClass yet and the current thread end up getting a stale value cachedDeclaringClass?

Reply:
The declaring class offset value can be saved to a local variable after `getUnsafe().objectFieldOffset`, and is only stored into `cachedDeclaringClassOffset` after `cachedDeclaringClass` is set.

Reply:
Multi-threaded, you can't rely on the order which variables are set unless barriers are used.

Reply:
I think here `getUnsafe().put...` ensures `cachedDeclaringClass` set atomically. Regardless `cachedDeclaringClassOffset` is set or not, `cachedDeclaringClass` value won't be stale. In worst scenarios, `cachedDeclaringClass` will be set more than once which is acceptable.
Any thing I missed?

Reply:
No, looking again the suggestion is good since the unsafe call provides the required barrier.